// react-native-paper/lib/commonjs/components/TextInput/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OUTLINE_MINIMIZED_LABEL_Y_OFFSET = exports.MIN_WIDTH = exports.MIN_DENSE_HEIGHT_WL = exports.MIN_DENSE_HEIGHT_OUTLINED = exports.MIN_DENSE_HEIGHT = exports.MINIMIZED_LABEL_Y_OFFSET = exports.MINIMIZED_LABEL_FONT_SIZE = exports.MD3_OUTLINED_INPUT_OFFSET = exports.MD3_MIN_HEIGHT = exports.MD3_LABEL_PADDING_TOP = exports.MD3_LABEL_PADDING_HORIZONTAL = exports.MD3_INPUT_PADDING_HORIZONTAL = exports.MD3_ICON_OFFSET = exports.MD3_FLAT_INPUT_OFFSET = exports.MD3_AFFIX_OFFSET = exports.MD3_ADORNMENT_OFFSET = exports.MD2_OUTLINED_INPUT_OFFSET = exports.MD2_MIN_HEIGHT = exports.MD2_LABEL_PADDING_TOP = exports.MD2_LABEL_PADDING_HORIZONTAL = exports.MD2_INPUT_PADDING_HORIZONTAL = exports.MD2_ICON_OFFSET = exports.MD2_FLAT_INPUT_OFFSET = exports.MD2_AFFIX_OFFSET = exports.MD2_ADORNMENT_OFFSET = exports.MAXIMIZED_LABEL_FONT_SIZE = exports.LABEL_WIGGLE_X_OFFSET = exports.LABEL_PADDING_TOP_DENSE = exports.LABEL_PADDING_TOP = exports.ICON_SIZE = exports.ADORNMENT_SIZE = void 0;
const MAXIMIZED_LABEL_FONT_SIZE = 16;
exports.MAXIMIZED_LABEL_FONT_SIZE = MAXIMIZED_LABEL_FONT_SIZE;
const MINIMIZED_LABEL_FONT_SIZE = 12;
exports.MINIMIZED_LABEL_FONT_SIZE = MINIMIZED_LABEL_FONT_SIZE;
const LABEL_WIGGLE_X_OFFSET = 4;
exports.LABEL_WIGGLE_X_OFFSET = LABEL_WIGGLE_X_OFFSET;
const ADORNMENT_SIZE = 24;
exports.ADORNMENT_SIZE = ADORNMENT_SIZE;
const MIN_WIDTH = 100;

//Text input affix offset
exports.MIN_WIDTH = MIN_WIDTH;
const MD2_AFFIX_OFFSET = 12;
exports.MD2_AFFIX_OFFSET = MD2_AFFIX_OFFSET;
const MD3_AFFIX_OFFSET = 16;

// Text input icon
exports.MD3_AFFIX_OFFSET = MD3_AFFIX_OFFSET;
const ICON_SIZE = 24;
exports.ICON_SIZE = ICON_SIZE;
const MD2_ICON_OFFSET = 12;
exports.MD2_ICON_OFFSET = MD2_ICON_OFFSET;
const MD3_ICON_OFFSET = 16;

// Text input common
exports.MD3_ICON_OFFSET = MD3_ICON_OFFSET;
const MD2_MIN_HEIGHT = 64;
exports.MD2_MIN_HEIGHT = MD2_MIN_HEIGHT;
const MD3_MIN_HEIGHT = 56;
exports.MD3_MIN_HEIGHT = MD3_MIN_HEIGHT;
const MD3_ADORNMENT_OFFSET = 16;
exports.MD3_ADORNMENT_OFFSET = MD3_ADORNMENT_OFFSET;
const MD2_ADORNMENT_OFFSET = 12;
exports.MD2_ADORNMENT_OFFSET = MD2_ADORNMENT_OFFSET;
const LABEL_PADDING_TOP_DENSE = 24;
exports.LABEL_PADDING_TOP_DENSE = LABEL_PADDING_TOP_DENSE;
const LABEL_PADDING_TOP = 8;

// Text input flat
exports.LABEL_PADDING_TOP = LABEL_PADDING_TOP;
const MD2_LABEL_PADDING_TOP = 30;
exports.MD2_LABEL_PADDING_TOP = MD2_LABEL_PADDING_TOP;
const MD3_LABEL_PADDING_TOP = 26;
exports.MD3_LABEL_PADDING_TOP = MD3_LABEL_PADDING_TOP;
const MD2_LABEL_PADDING_HORIZONTAL = 12;
exports.MD2_LABEL_PADDING_HORIZONTAL = MD2_LABEL_PADDING_HORIZONTAL;
const MD3_LABEL_PADDING_HORIZONTAL = 16;
exports.MD3_LABEL_PADDING_HORIZONTAL = MD3_LABEL_PADDING_HORIZONTAL;
const MD2_FLAT_INPUT_OFFSET = 8;
exports.MD2_FLAT_INPUT_OFFSET = MD2_FLAT_INPUT_OFFSET;
const MD3_FLAT_INPUT_OFFSET = 16;
exports.MD3_FLAT_INPUT_OFFSET = MD3_FLAT_INPUT_OFFSET;
const MINIMIZED_LABEL_Y_OFFSET = -18;
exports.MINIMIZED_LABEL_Y_OFFSET = MINIMIZED_LABEL_Y_OFFSET;
const MIN_DENSE_HEIGHT_WL = 52;
exports.MIN_DENSE_HEIGHT_WL = MIN_DENSE_HEIGHT_WL;
const MIN_DENSE_HEIGHT = 40;

// Text input outlined
exports.MIN_DENSE_HEIGHT = MIN_DENSE_HEIGHT;
const MD2_INPUT_PADDING_HORIZONTAL = 14;
exports.MD2_INPUT_PADDING_HORIZONTAL = MD2_INPUT_PADDING_HORIZONTAL;
const MD3_INPUT_PADDING_HORIZONTAL = 16;

// extra space to avoid overlapping input's text and icon
exports.MD3_INPUT_PADDING_HORIZONTAL = MD3_INPUT_PADDING_HORIZONTAL;
const MD2_OUTLINED_INPUT_OFFSET = 8;
exports.MD2_OUTLINED_INPUT_OFFSET = MD2_OUTLINED_INPUT_OFFSET;
const MD3_OUTLINED_INPUT_OFFSET = 16;
exports.MD3_OUTLINED_INPUT_OFFSET = MD3_OUTLINED_INPUT_OFFSET;
const OUTLINE_MINIMIZED_LABEL_Y_OFFSET = -6;
exports.OUTLINE_MINIMIZED_LABEL_Y_OFFSET = OUTLINE_MINIMIZED_LABEL_Y_OFFSET;
const MIN_DENSE_HEIGHT_OUTLINED = 48;
exports.MIN_DENSE_HEIGHT_OUTLINED = MIN_DENSE_HEIGHT_OUTLINED;
