// react-native-svg/lib/commonjs/fabric/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "RNSVGCircle", {
  enumerable: true,
  get: function () {
    return _CircleNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGClipPath", {
  enumerable: true,
  get: function () {
    return _ClipPathNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGDefs", {
  enumerable: true,
  get: function () {
    return _DefsNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGEllipse", {
  enumerable: true,
  get: function () {
    return _EllipseNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGForeignObject", {
  enumerable: true,
  get: function () {
    return _ForeignObjectNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGGroup", {
  enumerable: true,
  get: function () {
    return _GroupNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGImage", {
  enumerable: true,
  get: function () {
    return _ImageNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGLine", {
  enumerable: true,
  get: function () {
    return _LineNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGLinearGradient", {
  enumerable: true,
  get: function () {
    return _LinearGradientNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGMarker", {
  enumerable: true,
  get: function () {
    return _MarkerNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGMask", {
  enumerable: true,
  get: function () {
    return _MaskNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGPath", {
  enumerable: true,
  get: function () {
    return _PathNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGPattern", {
  enumerable: true,
  get: function () {
    return _PatternNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGRadialGradient", {
  enumerable: true,
  get: function () {
    return _RadialGradientNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGRect", {
  enumerable: true,
  get: function () {
    return _RectNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGSvgAndroid", {
  enumerable: true,
  get: function () {
    return _AndroidSvgViewNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGSvgIOS", {
  enumerable: true,
  get: function () {
    return _IOSSvgViewNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGSymbol", {
  enumerable: true,
  get: function () {
    return _SymbolNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGTSpan", {
  enumerable: true,
  get: function () {
    return _TSpanNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGText", {
  enumerable: true,
  get: function () {
    return _TextNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGTextPath", {
  enumerable: true,
  get: function () {
    return _TextPathNativeComponent.default;
  }
});
Object.defineProperty(exports, "RNSVGUse", {
  enumerable: true,
  get: function () {
    return _UseNativeComponent.default;
  }
});
var _CircleNativeComponent = _interopRequireDefault(require("./CircleNativeComponent"));
var _ClipPathNativeComponent = _interopRequireDefault(require("./ClipPathNativeComponent"));
var _DefsNativeComponent = _interopRequireDefault(require("./DefsNativeComponent"));
var _EllipseNativeComponent = _interopRequireDefault(require("./EllipseNativeComponent"));
var _ForeignObjectNativeComponent = _interopRequireDefault(require("./ForeignObjectNativeComponent"));
var _GroupNativeComponent = _interopRequireDefault(require("./GroupNativeComponent"));
var _ImageNativeComponent = _interopRequireDefault(require("./ImageNativeComponent"));
var _LinearGradientNativeComponent = _interopRequireDefault(require("./LinearGradientNativeComponent"));
var _LineNativeComponent = _interopRequireDefault(require("./LineNativeComponent"));
var _MarkerNativeComponent = _interopRequireDefault(require("./MarkerNativeComponent"));
var _MaskNativeComponent = _interopRequireDefault(require("./MaskNativeComponent"));
var _PathNativeComponent = _interopRequireDefault(require("./PathNativeComponent"));
var _PatternNativeComponent = _interopRequireDefault(require("./PatternNativeComponent"));
var _RadialGradientNativeComponent = _interopRequireDefault(require("./RadialGradientNativeComponent"));
var _RectNativeComponent = _interopRequireDefault(require("./RectNativeComponent"));
var _AndroidSvgViewNativeComponent = _interopRequireDefault(require("./AndroidSvgViewNativeComponent"));
var _IOSSvgViewNativeComponent = _interopRequireDefault(require("./IOSSvgViewNativeComponent"));
var _SymbolNativeComponent = _interopRequireDefault(require("./SymbolNativeComponent"));
var _TextNativeComponent = _interopRequireDefault(require("./TextNativeComponent"));
var _TextPathNativeComponent = _interopRequireDefault(require("./TextPathNativeComponent"));
var _TSpanNativeComponent = _interopRequireDefault(require("./TSpanNativeComponent"));
var _UseNativeComponent = _interopRequireDefault(require("./UseNativeComponent"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
