// react-native-svg/lib/commonjs/ReactNativeSVG.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Shape: true,
  Rect: true,
  Circle: true,
  Ellipse: true,
  Polygon: true,
  Polyline: true,
  Line: true,
  Svg: true,
  Path: true,
  G: true,
  Text: true,
  TSpan: true,
  TextPath: true,
  Use: true,
  Image: true,
  Symbol: true,
  Defs: true,
  LinearGradient: true,
  RadialGradient: true,
  Stop: true,
  ClipPath: true,
  Pattern: true,
  Mask: true,
  Marker: true,
  ForeignObject: true,
  parse: true,
  SvgAst: true,
  SvgFromUri: true,
  SvgFromXml: true,
  SvgUri: true,
  SvgXml: true,
  camelCase: true,
  err: true,
  fetchText: true,
  RNSVGCircle: true,
  RNSVGClipPath: true,
  RNSVGDefs: true,
  RNSVGEllipse: true,
  RNSVGForeignObject: true,
  RNSVGGroup: true,
  RNSVGImage: true,
  RNSVGLine: true,
  RNSVGLinearGradient: true,
  RNSVGMarker: true,
  RNSVGMask: true,
  RNSVGPath: true,
  RNSVGPattern: true,
  RNSVGRadialGradient: true,
  RNSVGRect: true,
  RNSVGSvgAndroid: true,
  RNSVGSvgIOS: true,
  RNSVGSymbol: true,
  RNSVGText: true,
  RNSVGTextPath: true,
  RNSVGTSpan: true,
  RNSVGUse: true,
  SvgCss: true,
  SvgCssUri: true,
  SvgWithCss: true,
  SvgWithCssUri: true,
  inlineStyles: true,
  LocalSvg: true,
  WithLocalSvg: true,
  loadLocalRawResource: true
};
Object.defineProperty(exports, "Circle", {
  enumerable: true,
  get: function () {
    return _Circle.default;
  }
});
Object.defineProperty(exports, "ClipPath", {
  enumerable: true,
  get: function () {
    return _ClipPath.default;
  }
});
Object.defineProperty(exports, "Defs", {
  enumerable: true,
  get: function () {
    return _Defs.default;
  }
});
Object.defineProperty(exports, "Ellipse", {
  enumerable: true,
  get: function () {
    return _Ellipse.default;
  }
});
Object.defineProperty(exports, "ForeignObject", {
  enumerable: true,
  get: function () {
    return _ForeignObject.default;
  }
});
Object.defineProperty(exports, "G", {
  enumerable: true,
  get: function () {
    return _G.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _Image.default;
  }
});
Object.defineProperty(exports, "Line", {
  enumerable: true,
  get: function () {
    return _Line.default;
  }
});
Object.defineProperty(exports, "LinearGradient", {
  enumerable: true,
  get: function () {
    return _LinearGradient.default;
  }
});
Object.defineProperty(exports, "LocalSvg", {
  enumerable: true,
  get: function () {
    return _deprecated.LocalSvg;
  }
});
Object.defineProperty(exports, "Marker", {
  enumerable: true,
  get: function () {
    return _Marker.default;
  }
});
Object.defineProperty(exports, "Mask", {
  enumerable: true,
  get: function () {
    return _Mask.default;
  }
});
Object.defineProperty(exports, "Path", {
  enumerable: true,
  get: function () {
    return _Path.default;
  }
});
Object.defineProperty(exports, "Pattern", {
  enumerable: true,
  get: function () {
    return _Pattern.default;
  }
});
Object.defineProperty(exports, "Polygon", {
  enumerable: true,
  get: function () {
    return _Polygon.default;
  }
});
Object.defineProperty(exports, "Polyline", {
  enumerable: true,
  get: function () {
    return _Polyline.default;
  }
});
Object.defineProperty(exports, "RNSVGCircle", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGCircle;
  }
});
Object.defineProperty(exports, "RNSVGClipPath", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGClipPath;
  }
});
Object.defineProperty(exports, "RNSVGDefs", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGDefs;
  }
});
Object.defineProperty(exports, "RNSVGEllipse", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGEllipse;
  }
});
Object.defineProperty(exports, "RNSVGForeignObject", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGForeignObject;
  }
});
Object.defineProperty(exports, "RNSVGGroup", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGGroup;
  }
});
Object.defineProperty(exports, "RNSVGImage", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGImage;
  }
});
Object.defineProperty(exports, "RNSVGLine", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGLine;
  }
});
Object.defineProperty(exports, "RNSVGLinearGradient", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGLinearGradient;
  }
});
Object.defineProperty(exports, "RNSVGMarker", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGMarker;
  }
});
Object.defineProperty(exports, "RNSVGMask", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGMask;
  }
});
Object.defineProperty(exports, "RNSVGPath", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGPath;
  }
});
Object.defineProperty(exports, "RNSVGPattern", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGPattern;
  }
});
Object.defineProperty(exports, "RNSVGRadialGradient", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGRadialGradient;
  }
});
Object.defineProperty(exports, "RNSVGRect", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGRect;
  }
});
Object.defineProperty(exports, "RNSVGSvgAndroid", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGSvgAndroid;
  }
});
Object.defineProperty(exports, "RNSVGSvgIOS", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGSvgIOS;
  }
});
Object.defineProperty(exports, "RNSVGSymbol", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGSymbol;
  }
});
Object.defineProperty(exports, "RNSVGTSpan", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGTSpan;
  }
});
Object.defineProperty(exports, "RNSVGText", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGText;
  }
});
Object.defineProperty(exports, "RNSVGTextPath", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGTextPath;
  }
});
Object.defineProperty(exports, "RNSVGUse", {
  enumerable: true,
  get: function () {
    return _fabric.RNSVGUse;
  }
});
Object.defineProperty(exports, "RadialGradient", {
  enumerable: true,
  get: function () {
    return _RadialGradient.default;
  }
});
Object.defineProperty(exports, "Rect", {
  enumerable: true,
  get: function () {
    return _Rect.default;
  }
});
Object.defineProperty(exports, "Shape", {
  enumerable: true,
  get: function () {
    return _Shape.default;
  }
});
Object.defineProperty(exports, "Stop", {
  enumerable: true,
  get: function () {
    return _Stop.default;
  }
});
Object.defineProperty(exports, "Svg", {
  enumerable: true,
  get: function () {
    return _Svg.default;
  }
});
Object.defineProperty(exports, "SvgAst", {
  enumerable: true,
  get: function () {
    return _xml.SvgAst;
  }
});
Object.defineProperty(exports, "SvgCss", {
  enumerable: true,
  get: function () {
    return _deprecated.SvgCss;
  }
});
Object.defineProperty(exports, "SvgCssUri", {
  enumerable: true,
  get: function () {
    return _deprecated.SvgCssUri;
  }
});
Object.defineProperty(exports, "SvgFromUri", {
  enumerable: true,
  get: function () {
    return _xml.SvgFromUri;
  }
});
Object.defineProperty(exports, "SvgFromXml", {
  enumerable: true,
  get: function () {
    return _xml.SvgFromXml;
  }
});
Object.defineProperty(exports, "SvgUri", {
  enumerable: true,
  get: function () {
    return _xml.SvgUri;
  }
});
Object.defineProperty(exports, "SvgWithCss", {
  enumerable: true,
  get: function () {
    return _deprecated.SvgWithCss;
  }
});
Object.defineProperty(exports, "SvgWithCssUri", {
  enumerable: true,
  get: function () {
    return _deprecated.SvgWithCssUri;
  }
});
Object.defineProperty(exports, "SvgXml", {
  enumerable: true,
  get: function () {
    return _xml.SvgXml;
  }
});
Object.defineProperty(exports, "Symbol", {
  enumerable: true,
  get: function () {
    return _Symbol.default;
  }
});
Object.defineProperty(exports, "TSpan", {
  enumerable: true,
  get: function () {
    return _TSpan.default;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _Text.default;
  }
});
Object.defineProperty(exports, "TextPath", {
  enumerable: true,
  get: function () {
    return _TextPath.default;
  }
});
Object.defineProperty(exports, "Use", {
  enumerable: true,
  get: function () {
    return _Use.default;
  }
});
Object.defineProperty(exports, "WithLocalSvg", {
  enumerable: true,
  get: function () {
    return _deprecated.WithLocalSvg;
  }
});
Object.defineProperty(exports, "camelCase", {
  enumerable: true,
  get: function () {
    return _xml.camelCase;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "err", {
  enumerable: true,
  get: function () {
    return _xml.err;
  }
});
Object.defineProperty(exports, "fetchText", {
  enumerable: true,
  get: function () {
    return _xml.fetchText;
  }
});
Object.defineProperty(exports, "inlineStyles", {
  enumerable: true,
  get: function () {
    return _deprecated.inlineStyles;
  }
});
Object.defineProperty(exports, "loadLocalRawResource", {
  enumerable: true,
  get: function () {
    return _deprecated.loadLocalRawResource;
  }
});
Object.defineProperty(exports, "parse", {
  enumerable: true,
  get: function () {
    return _xml.parse;
  }
});
var _Shape = _interopRequireDefault(require("./elements/Shape"));
var _Rect = _interopRequireDefault(require("./elements/Rect"));
var _Circle = _interopRequireDefault(require("./elements/Circle"));
var _Ellipse = _interopRequireDefault(require("./elements/Ellipse"));
var _Polygon = _interopRequireDefault(require("./elements/Polygon"));
var _Polyline = _interopRequireDefault(require("./elements/Polyline"));
var _Line = _interopRequireDefault(require("./elements/Line"));
var _Svg = _interopRequireDefault(require("./elements/Svg"));
var _Path = _interopRequireDefault(require("./elements/Path"));
var _G = _interopRequireDefault(require("./elements/G"));
var _Text = _interopRequireDefault(require("./elements/Text"));
var _TSpan = _interopRequireDefault(require("./elements/TSpan"));
var _TextPath = _interopRequireDefault(require("./elements/TextPath"));
var _Use = _interopRequireDefault(require("./elements/Use"));
var _Image = _interopRequireDefault(require("./elements/Image"));
var _Symbol = _interopRequireDefault(require("./elements/Symbol"));
var _Defs = _interopRequireDefault(require("./elements/Defs"));
var _LinearGradient = _interopRequireDefault(require("./elements/LinearGradient"));
var _RadialGradient = _interopRequireDefault(require("./elements/RadialGradient"));
var _Stop = _interopRequireDefault(require("./elements/Stop"));
var _ClipPath = _interopRequireDefault(require("./elements/ClipPath"));
var _Pattern = _interopRequireDefault(require("./elements/Pattern"));
var _Mask = _interopRequireDefault(require("./elements/Mask"));
var _Marker = _interopRequireDefault(require("./elements/Marker"));
var _ForeignObject = _interopRequireDefault(require("./elements/ForeignObject"));
var _xml = require("./xml");
var _fabric = require("./fabric");
var _deprecated = require("./deprecated");
var _types = require("./lib/extract/types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = _Svg.default;
exports.default = _default;
